File: js/lib/WowCbor.js
'use strict';
const { decode, encode } = require('cbor-x');

/**
 * Post-process a decoded CBOR value:
 * 1. CBOR byte strings (Uint8Array / Buffer) → UTF-8 text strings
 * 2. Sequential 1-based integer-keyed maps → arrays (Lua table convention)
 * 3. Maps returned as Map objects are also handled
 */
function postProcess(val) {
    if (val instanceof Uint8Array || Buffer.isBuffer(val)) {
        return Buffer.from(val).toString('utf8');
    }
    if (Array.isArray(val)) {
        return val.map(postProcess);
    }
    if (val instanceof Map) {
        // cbor-x may return Map for non-string-keyed CBOR maps
        const obj = {};
        for (const [k, v] of val) {
            obj[k] = postProcess(v);
        }
        return applyArrayDetection(obj);
    }
    if (val !== null && typeof val === 'object') {
        const result = {};
        for (const [k, v] of Object.entries(val)) {
            result[k] = postProcess(v);
        }
        return applyArrayDetection(result);
    }
    return val;
}

function applyArrayDetection(obj) {
    const keys = Object.keys(obj);
    if (keys.length === 0) return obj;
    const nums = keys.map(k => Number(k));
    if (!nums.every(n => Number.isInteger(n) && n > 0)) return obj;
    const sorted = [...nums].sort((a, b) => a - b);
    if (sorted.every((n, i) => n === i + 1)) {
        return sorted.map(n => obj[String(n)]);
    }
    return obj;
}

class WowCbor {
    /**
     * Decode CBOR bytes to a JS value, with WoW-specific post-processing.
     * @param {Buffer|Uint8Array} bytes
     * @returns {*}
     */
    static decode(bytes) {
        const raw = decode(Buffer.isBuffer(bytes) ? bytes : Buffer.from(bytes));
        return postProcess(raw);
    }

    /**
     * Encode a JS value to CBOR bytes.
     * @param {*} data
     * @returns {Buffer}
     */
    static encode(data) {
        const raw = encode(data);
        return Buffer.isBuffer(raw) ? raw : Buffer.from(raw);
    }
}

module.exports = WowCbor;

File: ruby/wow_cbor.rb
# frozen_string_literal: true

require 'cbor'

# WowCbor wraps the CBOR library with WoW-specific post-processing:
# - CBOR byte strings (binary Ruby strings) are decoded to UTF-8 text
# - Sequential 1-based integer-keyed maps are converted to arrays (Lua table convention)
class WowCbor
  class << self
    # Decode CBOR bytes to a Ruby value with WoW-specific post-processing.
    def decode(bytes)
      raw = CBOR.decode(bytes)
      post_process(raw)
    end

    # Encode a Ruby value to CBOR bytes.
    def encode(data)
      data.to_cbor
    end

    private

    def post_process(val)
      case val
      when String
        # CBOR byte strings come back as binary-encoded Ruby strings
        if val.encoding == Encoding::BINARY || val.encoding == Encoding::ASCII_8BIT
          val.encode('UTF-8', invalid: :replace, undef: :replace)
        else
          val
        end
      when Array
        val.map { |v| post_process(v) }
      when Hash
        result = val.transform_values { |v| post_process(v) }
        apply_array_detection(result)
      else
        val
      end
    end

    def apply_array_detection(hash)
      keys = hash.keys
      return hash if keys.empty?
      return hash unless keys.all? { |k| k.is_a?(Integer) && k > 0 }

      sorted = keys.sort
      return hash unless sorted == (1..keys.size).to_a

      sorted.map { |k| hash[k] }
    end
  end
end
